pkg/metrics/metric/metric.go

```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package metric
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// WithMetadata is the interface implemented by any metric defined in this package. These typically embed existing
+// prometheus metric types and add additional metadata to the definitions which can be used to generate proper metric
+// documentation from de definitions. In addition, these metrics have the concept of being enabled or disabled which
+// is used in place of conditional registration so all metric types can always be registered for the purposes of
+// documenting them.
+type WithMetadata interface {
+	IsEnabled() bool
+	SetEnabled(bool)
+	Opts() Opts
+}
+
+// metric is a "base" structure which can be embedded to provide common functionality.
+type metric struct {
+	enabled bool
+	opts    Opts
+}
+
+func (b *metric) IsEnabled() bool {
+	return b.enabled
+}
+
+func (b *metric) SetEnabled(e bool) {
+	b.enabled = e
+}
+
+// Vec is a generic type to describe the vectorized version of another metric type, for example Vec[Counter] would be
+// our version of a prometheus.CounterVec.
+type Vec[T any] interface {
+	prometheus.Collector
+	WithMetadata
+
+	CurryWith(labels prometheus.Labels) (Vec[T], error)
+	GetMetricWith(labels prometheus.Labels) (T, error)
+	GetMetricWithLabelValues(lvs ...string) (T, error)
+	With(labels prometheus.Labels) T
+	WithLabelValues(lvs ...string) T
+}
+
+// DeletableVec is a generic type to describe a vectorized version of another metric type, like Vec[T], but with the
+// additional ability to remove labels without re-creating the metric.
+type DeletableVec[T any] interface {
+	Vec[T]
+	Delete(labels prometheus.Labels) bool
+	DeleteLabelValues(lvs ...string) bool
+	DeletePartialMatch(labels prometheus.Labels) int
+	Reset()
+}
+
+// Opts are a modified and extended version of the prometheus.Opts
```

Comment:
There are two reasons for not doing so. First is that I expect to change the types of `Subsystem` in a later PR which will make embedding impossible. Secondly, making literals with embedded objects is a pain:

```
opt := NoEmbeds{
  Name: "",
  Namespace: "",
  EnabledByDefault: true,
}

opt2 := WithEmbeds{
  opts: prometheus.Opts{
    Name: "",
    Namespace: "",
  },
  EnableByDefault: true,
}
```

Reply:
Ack re the first reason. The reason why I asked is because we'd get future changes in prometheus.Opts for free if we embed; as is the maintenance is potentially increased (?).